Add Sts credentials provider

cmsd2 opened the pull request:
fyi This is blocked waiting on IAM support to be merged in.

I've added a credentials provider that uses the STS to acquire session tokens possibly using a role arn.

Example usage here: [https://github.com/cmsd2/stscli](https://github.com/cmsd2/stscli)

I also found I had to modify the request signing to use a stricter percent encoding for the canonical query string.

matthewkmayer commented:
@cmsd2 how can we help pick this back up now that IAM is merged in?

cmsd2 commented:
I've rebased it so you can see what I've added.
The StsProvider is designed to sit on top of a ProfileProvider, and uses ~/.aws/config in addition to ~/.aws/credentials
We use iam roles to make switching between accounts easier so I've written it with that in mind.
We don't use iam instance roles much, but that'd be a good usecase.
MFA is really needed, but I haven't tackled that yet.

I suppose for an initial pass, is the structure of the providers ok?

matthewkmayer commented on src/credential.rs:
```diff
@@ -117,6 +131,7 @@ pub trait ProvideAwsCredentials {
 }
 
 /// Provides AWS credentials from environment variables.
+#[derive(Debug, Clone)]
```
I see this is used for the threadsafe bits below in this file.  Is there a way we can get around using `derive (Debug, Clone)` ?  Derives are known to increase compile time and we should be sensitive to that. https://github.com/rusoto/rusoto/issues/364 is the parent issue for it.

cmsd2 replied:
ah. well i can remove the Debug impls.
the Clone is used to allow stacking providers -- the StsProvider creates an StsClient to do its work. Creating a client consumes the provider so I added the ability to clone the providers that rusoto provides to allow them all to be used with this system.

matthewkmayer commented on src/credential.rs:
```diff
@@ -306,7 +355,309 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
+    }
+}
+
+#[derive(Debug, Clone)]
```
Same as above, be nice to avoid the derives if possible.

matthewkmayer commented on src/credential.rs:
```diff
@@ -306,7 +355,309 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
+    }
+}
+
+#[derive(Debug, Clone)]
```
Same `derive` comment. 😄

matthewkmayer commented on src/credential.rs:
```diff
@@ -306,7 +355,309 @@ fn parse_credentials_file(file_path: &Path) -> Result<HashMap<String, AwsCredent
+}
+
+#[cfg(feature = "sts")]
+mod sts {
```
This `credentials.rs` file is getting a bit big.  This area seems like a good place to split into its own file if possible.

matthewkmayer commented:
@cmsd2 what can we do to help?

cmsd2 commented:
obviously i need to bring this branch up-to-date so it merges cleanly.
if you're happy with the splitting up of credentials.rs that we did then that's fine now too.

i think though that i didn't pay enough attention to what the other sdks do.
the ergonomics of this api might be a little different to e.g. the java sdk,
and it would be good to ensure we can handle the same combinations of env vars and config files to avoid surprises.
if you're looking at a 1.0 release then i'd want to think hard about how forward compatible credential.rs is wrt sts.

i don't suppose anyone else around here uses sts?
has anyone tried something exotic like using https://github.com/apex/apex to run a rust process inside a lambda and checking that we can use the execution role to call aws services?

cmsd2 commented:
ok. new approach.
i've removed all the slightly magic stuff to do with profiles and followed the layout of the java sdk.
it means you have multiple providers, one for each usecase.

i also found that with the openid connect provider it needs somewhere to save the subject claim. the java sdk providers save results like that in each provider. i added a map to awscredentials as a place to save that to avoid putting interior mutability in the provider.
one niggle: since providers are owned by each client it's not easy to poke the provider after handing it to the client. what would be a good way of getting at the claims in the creds that the client gets from the provider?

i still haven't got a good way to do mfa. i think it would be normal to try e.g. assume_role without mfa, detect the access denied, prompt the user for mfa and retry.
that would require a more sophisticated wrapper layer between client and provider, with app supplied callbacks. that will have to wait.

matthewkmayer commented:
Thanks for sticking with this PR, it's been a rough road, but we'll get it done.

I like how we've split the credentials crate.  You're totally right we need to take a hard look at it to see if we can improve ergonomics.  Perhaps get all services codegened and basic integration tests running, then do more refactors?

It's also a good test to see if we can use `apex` to get a Rusoto-using Rust program working as expected.  I'll make a new ticket for that, hopefully someone can pick that up.

$DAY_JOB has a scenario using IAM role delegation between accounts so I may get some STS exposure through that!

Finally, we may need to rethink service clients owning credential providers as you suggest.  Perhaps some spelunking into Java or .NET SDKs can help us.

adimarco commented:
Hey @cmsd2 - I'm wrapping up a pretty significant refactor of the `query` protocol codegen, of which STS is a member.  That should land in the next day or two.  I'll make sure that STS gets properly generated as part of that, and you can take it from there with whatever modifications need to be made to the credentials provider.  Sound good?

cmsd2 commented:
Hi @adimarco 
Thanks, I'll keep an eye on your changes.
I've been able to validate that the code in my brunch is functional at least for the use cases I tested -- I updated my stscli tool to cope.
If your changes answer the last remaining usability questions then that's potentially very interesting.

matthewkmayer commented:
@cmsd2 the big codegen refactor has landed, we're ready to get back at this.  Let's make a list of questions to answer and work to do to land this PR.

If there's a deep discussion to do, perhaps a Hangout would work.  If you think that will help get this all done, my email address is in [the Cargo.toml file](https://github.com/rusoto/rusoto/blob/master/Cargo.toml#L5).

cmsd2 commented:
I had a go with mfa and assume_role today.
It works fine without any changes. The application just needs to know in advance that it needs to supply an mfa token when assuming the role.
The aws api doesn't seem to give any special consideration for nice challenge and response flows, so rusoto probably shouldn't either. It just leaves some work for the end application.

Regarding allowing apps to inspect credentials so they can access the claims from federated access:
Would it make sense to change the AutoRefreshingProviderSync to use an Arc<Mutex>?
That would allow an app to retain a clone of the provider and get access to the credentials outside of the client(s), which would take ownership of the other clone(s).

matthewkmayer commented:
I thought the updates made in the source repo would remove the merge conflicts.  Looks like the branch in your repo doesn't have some updates yet.  Can you either rebase onto current master or merge master into your branch to fix those?

Thanks!

cmsd2 commented:
rebased.

cmsd2 commented:
let me know if there's anything more I can do.

matthewkmayer commented:
I'm back able to look at this PR.  First I'm whittling away at other, smaller PRs for a morale boost, then I'm going to give this a good review.

Thank you again for sticking with it, it's appreciated.

matthewkmayer commented on Cargo.toml:
```diff
@@ -107,6 +107,7 @@ all = [
 	   	"sqs",
 	   	"ssm",
 	   	"storagegateway",
+                "sts",
```
Nit: spacing is off.

matthewkmayer commented on src/sts.rs:
```diff
@@ -0,0 +1,269 @@
+//! The AWS STS API.
+
```
Let's add a line in here to make it easier to find the main client class.  This is https://github.com/rusoto/rusoto/issues/519 and there are examples in https://github.com/rusoto/rusoto/pull/544 .  A gotcha is the client struct URL is case sensitive when it's deployed to the API reference page on Github pages (https://rusoto.github.io/rusoto/rusoto/).

matthewkmayer commented:
These changes look good to me.  There's still the unanswered question around `AutoRefreshingProviderSync`.  Can we address that in a future PR so we can get this behavior in?  I can make a new issue for that if so.

Of course the change for getting request bodies as `Vec<u8>` caused merge conflict, and this may require tweaking to ensure the body is converted from the byte vector to something that can be converted to a String.

cmsd2 commented:
merged and patched.
how does that look?

matthewkmayer commented:
This is on deck for me to review in the next day or two.  I will not be merging anything else until then because I'm tired of constantly introducing merge conflicts in this PR. 😄

matthewkmayer commented on credential/src/lib.rs:
```diff
@@ -123,6 +137,12 @@ pub trait ProvideAwsCredentials {
     fn credentials(&self) -> Result<AwsCredentials, CredentialsError>;
 }
 
+impl ProvideAwsCredentials for AwsCredentials {
```
I'm not clear on why this is needed.  Can I see an example of this being used?

cmsd2 replied:
I can't find where this is used either. I'll remove it.

cmsd2 replied:
it was used in tests/sts.rs but only in passing and i'll remove that.
it can be used as a way of injecting basic credentials into a client without a further wrapper type that itself implements ProvideAwsCredentials.

matthewkmayer commented on src/sts.rs:
```diff
@@ -0,0 +1,278 @@
+//! The AWS STS API.
+//!
+//! If you're using the service, you're probably looking for [StsClient](struct.StsClient.html).
+#![allow(unused_imports)]
```
I'd like to remove this so we only import what we need.  When I remove the line there's four unused imports.  I'll look at it some more to see what it takes to remove this and still have the `sts` feature compile and work.

matthewkmayer replied:
I removed the imports in question and all STS tests pass.  However they aren't as thorough as actually using the service.  Can you make those changes in your fork and verify it works with your STS project?

cmsd2 replied:
Yes, this shouldn't be there and won't cause any harm to clean up.
I have a hunch that there are some combinations of features that cause things to be imported but then not used and then trip an unused import error. I suppose we'll fix that as and when.

matthewkmayer commented on src/sts.rs:
```diff
@@ -0,0 +1,278 @@
+
+include!(concat!(env!("OUT_DIR"), "/sts.rs"));
+
+pub use self::credential::{
```
Are the contents of this file needed because our current codegen makes the client for STS but none of the Rusoto credential required behavior?  I believe that's the reason but would like confirmation so I better understand it.  And if that's the case, let's add a comment in the code to that effect.

cmsd2 replied:
The generated sts code gives us the StsClient and all the requests, responses and errors. As you'd expect.
What we don't get is all the glue to create providers that the rusoto-credential module can use with it's provider chain.

This file implemnts all that in a private module (self::credential), then rexports selected symbols. (That's just a defensive measure to avoid colliding with anything imported from the generated sts module.

matthewkmayer commented on src/sts.rs:
```diff
@@ -0,0 +1,278 @@
+    pub const DEFAULT_DURATION_SECONDS: i32 = 3600;
+    pub const DEFAULT_ROLE_DURATION_SECONDS: i32 = 900;
+
+    pub trait NewAwsCredsForStsCreds {
```
I'd love to see comments on anything publicly exported so it'll show up in our API docs.

cmsd2 replied:
These are public constants in a private module. They are not re-exported.
The default durations are lifted from the java sdk, but also match the published defaults on the STS API itself.
Those defaults are documented in the methods on the generated StsClient.

cmsd2 replied:
I have started working on a change to properly document all the public types.

matthewkmayer commented:
👍 thank you for your work on this.  Aside from the few comments I made it looks good and we're almost there!

matthewkmayer commented:
🎉 Looks good to me!  Anything else to add or can we merge this in?

cmsd2 commented:
nothing else to add. ship it!
